### DISCUSSION TITLE
Should `File::hasSameHash` use `hash_equals`?

### DISCUSSION THREAD

**Discussion by simonhamp** [laravel/framework#49718] [Q&A] [answered]
**This is already 'fixed' in https://github.com/laravel/framework/pull/49721**

---

**Please don't take this as if I know what I'm talking about. This is an open question for discussion.**

I just came across [`File::hasSameHash()`](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Filesystem/Filesystem.php#L545-L550) and thought "hang on, didn't @valorin say [we shouldn't compare hashes with `===`](https://securinglaravel.com/p/security-tip-compare-keys-with-hash_equals)?"

While it seems infeasible that a file comparison would, _in practice_, be vulnerable to a timing attack (due to the sheer amount of binary data and therefore the processing time/cost required to enumerate a final hash), right now it's _technically possible_ to do if you discover that someone is using `File::hasSameHash()`.

## Implications
Especially in the context of this being an `md5` hash comparison (y'know, [_that_ MD5](https://en.wikipedia.org/wiki/MD5#:~:text=One%20basic%20requirement,for%20further%20use%22.)), it's entirely possible (however unlikely) that someone could use the enumerate-able `hasSameHash` method to craft a file that ends up with the same hash as a known safe file hash, and thus **send unexpected/faulty data into your system**.

Assuming the data in the uploaded file itself will be inert due to the 'random' nature of the enumeration (it will likely just be a mess of bytes), the main consequence of this "vulnerability" will be **errors and exceptions from trying to read said file** after it's passed your hash comparison, as it won't contain what you thought it did.

---

I haven't seen this happen in practice. I haven't got a proof of concept or anything like that. This is all just hypothetical/theoretical.

A quick GitHub code search doesn't reveal any repos actually _calling_ `hasSameHash` (although there seem to be [quite a few replicas](https://github.com/search?type=code&auto_enroll=true&q=hasSameHash+language%3APHP&l=PHP) of this implementation).

I'm also not going to pretend that I know what I'm talking about - I'm no infosec expert. I am just curious—having come across this just now—to see what other, more knowledgeable people think.

If this is indeed something that needs to be resolved, I will happily create the PR to fix - seems like a "good first issue".

**Comment by valorin** [comment, marked as answer]
Yep, it should definitely be using `hash_equals()` in there to compare the hashes. 👍

As you said, hash timing attacks are technically possible, but incredibly difficult to conduct. However, that's not a good reason to write weak code when we have a secure alternative that we can easily use instead.

Oh and I love that you're referencing Securing Laravel. 🥰

**Reply by simonhamp** [reply]
Thanks for chiming in on this so rapidly! Glad to know that I'm not barking mad 😅

**Comment by vlakoff** [comment]
See https://github.com/laravel/framework/pull/49721#issuecomment-2667240168:

> Note that the order of parameters in [`hash_equals()`](https://www.php.net/manual/en/function.hash-equals.php) matters.
> 
> The first parameter is the only one protected from guessing attempts.
